ethers-solc/src/artifacts/mod.rs

```diff
@@ -52,13 +52,37 @@ pub struct CompilerInput {
 
 impl CompilerInput {
     /// Reads all contracts found under the path
-    pub fn new(path: impl AsRef<Path>) -> Result<Self, SolcIoError> {
+    pub fn new(path: impl AsRef<Path>) -> Result<Vec<Self>, SolcIoError> {
         Source::read_all_from(path.as_ref()).map(Self::with_sources)
     }
 
     /// Creates a new Compiler input with default settings and the given sources
-    pub fn with_sources(sources: Sources) -> Self {
-        Self { language: "Solidity".to_string(), sources, settings: Default::default() }
+    pub fn with_sources(sources: Sources) -> Vec<Self> {
```

Comment:
Since this returns at most `vec![sol, yul]`
can we add something like this perhaps

```rust
enum ProjectCompilerInput {
  Solidity(CompilerInput),
  Yul(..),
  Mixed(..,..)
}
```
?

Reply:
I think the `ProjectCompilerInput` abstraction is correct right now - i.e. it's the unit that get's fed into solc compiler. The `with_sources` method though doesn't work because if sources have different types that means that you'll have 2 compiler inputs the needs to be fed into solc sequentially. So to me the signature looks good. Plus the vec is extensible for other sources (if any :))

To be 100% clean we could update the `new` method so it's just a plain constructor taking language as an input. For that method it's really uncommon to return a `Vec`. I just left it as it is for now because it could be used in many places.

Wdyt?

Reply:
makes sense, however working with a `Vec<CompilerInput>` can be a bit annoying.
Should be fine for now, but I think we can make this more ergonomic with `ProjectCompilerInput` which requires some refactoring, which we should leave for a follow up.

Reply:
I am OK with this as-is, seems good enough for now!